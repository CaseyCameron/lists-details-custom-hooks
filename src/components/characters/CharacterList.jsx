import React from 'react';
import { Link } from 'react-router-dom';
import Character from './Character';
import { useCharacters } from './hooks/useCharacters';

const CharacterList = ({ characters }) => {
  const [page, setPage] = useState(1)
  const { characters, loading } = useCharacters(page);
  if (loading) return <h1>Loading...</h1>;

  const characterElements = characters.map(character => (
    <li key={character._id}>
      <Link to={`/${character.id}`}>
        <Character {...character} />
      </Link>
    </li>
  ));

  return (
    <>
      <button
        disabled={page <= 1}
        oncClick={() => setPage((prevPage) => prevPage - 1)}
      >
        &lt;
      </button>
      <button onClick={() => setPage((prevPage) => prevPage + 1)}>&gt;</button>
      <ul>{characterElements}</ul>;
    </>
  );
};

export default CharacterList;