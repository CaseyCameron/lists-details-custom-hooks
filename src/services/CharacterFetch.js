export const fetchCharacters = async (page) => {
  const res = await fetch(
    `https://last-airbender-api.herokuapp.com/api/v1/characters?perPage=10&page=${page}`
  );
  const characters = await res.json();

  return characters;
};

export const fetchCharacterById = async (id) => {
  const res = await fetch(`https://last-airbender-api.herokuapp.com/api/v1/characters/${id}`);

  return res.json();
};