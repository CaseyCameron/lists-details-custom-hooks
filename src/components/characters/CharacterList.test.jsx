import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { AvatarCharacters } from '../../services/MockData';
import CharacterList from './CharacterList';

const server = setupServer(
  rest.get('https://last-airbender-api.herokuapp.com/api/v1/characters?perPage=10', (req, res, ctx) => {
    return res(ctx.json(AvatarCharacters))
  })
);

describe('Avatar characters list', () => {
  beforeAll(() => server.listen());
  afterAll(() => server.close());

  it('Renders a list of avatar characters', async () => {
    render(
      <MemoryRouter>
        <CharacterList />
      </MemoryRouter>
    );

    screen.getByText('Loading...');

    return waitFor(() => {
      const characters = screen.getByRole('list');
      expect(characters).not.toBeEmptyDOMElement();
      expect(characters).toMatchSnapshot();
    })
  });
});