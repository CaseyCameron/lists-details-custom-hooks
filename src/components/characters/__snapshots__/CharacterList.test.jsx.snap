// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar characters list Renders a list of avatar characters 1`] = `
<ul>
  <li>
    <a
      href="/5cf5679a915ecad153ab68d1"
    >
      <div
        aria-label="character"
      >
        <figure>
          <figcaption>
            <p>
              Analay
            </p>
          </figcaption>
          <img
            alt="Analay"
            src="https://vignette.wikia.nocookie.net/avatar/images/c/c3/Analay.png/revision/latest?cb=20140430182201"
          />
        </figure>
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5cf5679a915ecad153ab68d5"
    >
      <div
        aria-label="character"
      >
        <figure>
          <figcaption>
            <p>
              Arik
            </p>
          </figcaption>
          <img
            alt="Arik"
            src="https://vignette.wikia.nocookie.net/avatar/images/6/6f/Arik.png/revision/latest?cb=20140808184002"
          />
        </figure>
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5cf5679a915ecad153ab68e2"
    >
      <div
        aria-label="character"
      >
        <figure>
          <figcaption>
            <p>
              Baatar Jr.
            </p>
          </figcaption>
          <img
            alt="Baatar Jr."
            src="https://vignette.wikia.nocookie.net/avatar/images/4/45/Baatar_Jr..png/revision/latest?cb=20150409193510"
          />
        </figure>
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5cf5679a915ecad153ab68e9"
    >
      <div
        aria-label="character"
      >
        <figure>
          <figcaption>
            <p>
              Bolin
            </p>
          </figcaption>
          <img
            alt="Bolin"
            src="https://vignette.wikia.nocookie.net/avatar/images/d/d4/Bolin.png/revision/latest?cb=20150406232314"
          />
        </figure>
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5cf5679a915ecad153ab68fd"
    >
      <div
        aria-label="character"
      >
        <figure>
          <figcaption>
            <p>
              Chan (Fire Nation admiral)
            </p>
          </figcaption>
          <img
            alt="Chan (Fire Nation admiral)"
          />
        </figure>
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5cf5679a915ecad153ab6900"
    >
      <div
        aria-label="character"
      >
        <figure>
          <figcaption>
            <p>
              Chit Sang
            </p>
          </figcaption>
          <img
            alt="Chit Sang"
            src="https://vignette.wikia.nocookie.net/avatar/images/8/81/Chit_Sang.png/revision/latest?cb=20121021000707"
          />
        </figure>
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5cf5679a915ecad153ab68d7"
    >
      <div
        aria-label="character"
      >
        <figure>
          <figcaption>
            <p>
              Asami Sato
            </p>
          </figcaption>
          <img
            alt="Asami Sato"
            src="https://vignette.wikia.nocookie.net/avatar/images/b/b6/Asami_Sato.png/revision/latest?cb=20150407110150"
          />
        </figure>
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5cf5679a915ecad153ab68f0"
    >
      <div
        aria-label="character"
      >
        <figure>
          <figcaption>
            <p>
              Bum-Ju
            </p>
          </figcaption>
          <img
            alt="Bum-Ju"
            src="https://vignette.wikia.nocookie.net/avatar/images/b/b6/Bum-Ju.png/revision/latest?cb=20140730103156"
          />
        </figure>
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5cf5679a915ecad153ab68f5"
    >
      <div
        aria-label="character"
      >
        <figure>
          <figcaption>
            <p>
              Butler
            </p>
          </figcaption>
          <img
            alt="Butler"
            src="https://vignette.wikia.nocookie.net/avatar/images/8/82/Butler.png/revision/latest?cb=20121107115044"
          />
        </figure>
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5cf5679a915ecad153ab68ed"
    >
      <div
        aria-label="character"
      >
        <figure>
          <figcaption>
            <p>
              Broadsword man
            </p>
          </figcaption>
          <img
            alt="Broadsword man"
            src="https://vignette.wikia.nocookie.net/avatar/images/9/98/Broadsword_man.png/revision/latest?cb=20130630171553"
          />
        </figure>
      </div>
    </a>
  </li>
</ul>
`;
